pkg/bpmn/engine.go

```diff
@@ -631,20 +631,27 @@ func (engine *Engine) handleActivity(ctx context.Context, batch storage.Batch, i
 	var activityResult runtime.ActivityState
 	var err error
 
-	switch element := activity.Element().(type) {
-	case *bpmn20.TServiceTask:
-		activityResult, err = engine.createInternalTask(ctx, batch, instance, element, currentToken)
-	case *bpmn20.TSendTask:
-		activityResult, err = engine.createInternalTask(ctx, batch, instance, element, currentToken)
-	case *bpmn20.TUserTask:
-		activityResult, err = engine.createUserTask(ctx, batch, instance, element, currentToken)
-	case *bpmn20.TCallActivity:
-		activityResult, err = engine.createCallActivity(ctx, batch, instance, element, currentToken)
-		// we created process instance and its running in separate goroutine
-	case *bpmn20.TBusinessRuleTask:
-		activityResult, err = engine.createBusinessRuleTask(ctx, batch, instance, element, currentToken)
-	default:
-		return nil, fmt.Errorf("failed to process %s %d: %w", element.GetType(), activity.GetKey(), errors.New("unsupported activity"))
+	activityElement := engine.castToTActivity(element)
+
+	mi := activityElement.MultiInstanceLoopCharacteristics
+	isMultiInstance := mi.LoopCharacteristics.InputCollection != ""
+
+	// multi-instance
+	if isMultiInstance {
+		err = engine.initMultiInstances(ctx, instance, activityElement, &currentToken, batch, activity)
+		if err != nil {
+			return []runtime.ExecutionToken{currentToken}, err
+		}
+
+		// Create boundary event subscriptions for the multi-instance activity
+		err := createBoundaryEventSubscriptions(ctx, engine, batch, currentToken, instance, activity.Element())
+		if err != nil {
+			return nil, fmt.Errorf("failed to process boundary events for multi-instance %s %d: %w", element.GetType(), activity.GetKey(), err)
+		}
+
+		return []runtime.ExecutionToken{currentToken}, nil
+	} else {
```

Comment:
why is this in else branch if we are returning from if above? we should be able to get rid of the else statement and just continue